db/dcrpg/pgblockchain.go

```diff
@@ -4885,6 +4887,151 @@ func (pgb *ChainDB) GetTrimmedTransaction(txid *chainhash.Hash) *apitypes.Trimme
 	}
 }
 
+// GetAtomicSwapInfo gets a *dbtypes.AtomicSwapInfo for a given transaction ID,
+// if the transaction exists and represents an atomic swap, i.e.
+// - passed txid represents an atomic swap redemption/refund or
+// - one of the transaction's outputs was spent in a swap redempton/refund tx.
+// Returns nil if the passed txid does not represent an atomic swap (does not
+// match any of the aforementioned conditions).
+// NOTE: both return values may be nil if the txid is a valid transaction but
+// does not represent an atomic swap.
+func (pgb *ChainDB) GetAtomicSwapInfo(txid string) (*dbtypes.AtomicSwapInfo, error) {
+	swapInfo, err := pgb.GetAtomicSwapInfoFromRedemptionTx(txid)
+	if swapInfo != nil || err != nil {
+		return swapInfo, err
+	}
+
+	spendingTxIDs, _, _, err := pgb.SpendingTransactions(txid)
+	if err != nil {
+		log.Errorf("GetAtomicSwapInfo: retrieve spending txs error: %s", err)
+		return nil, fmt.Errorf("error checking if txid represents an atomic swap contract")
+	}
+
+	// Check this tx's outputs to determine if any was spent in a swap redemption tx.
+	for _, spendingTxID := range spendingTxIDs {
+		swapInfo, err = pgb.GetAtomicSwapInfoFromRedemptionTx(spendingTxID)
+		if swapInfo != nil || err != nil {
+			return swapInfo, err
+		}
+	}
+
+	return nil, nil
+}
+
+// GetAtomicSwapInfoFromRedemptionTx gets a *dbtypes.AtomicSwapInfo for a given
+// transaction ID, if the transaction exists and represents an atomic swap
+// redemption/refund.
+// Returns nil if the passed txid does not represent a swap redemption/refund.
+// NOTE: both return values may be nil if the txid is a valid transaction but
+// does not represent an atomic swap redemption/refund.
+func (pgb *ChainDB) GetAtomicSwapInfoFromRedemptionTx(txid string) (*dbtypes.AtomicSwapInfo, error) {
```

Comment:
The meat of `GetAtomicSwapInfoFromRedemptionTx` really does not need to be a `ChainDB` method since it's only using `getRawAPITransaction` and that just uses `rpcutils.APITransaction`.  This method can still exist, perhaps just wrapping an `rpcutils` function of the same name.  Having it in `rpcutils` would make it more broadly accessible, and more easily testable.

Reply:
But I'd eliminate the `Get` prefix (we shouldn't have such names although we have many) and go with `AtomicSwapInfoFromRedemption` or similar.

Reply:
Had similar thoughts re the method name but just wanted to follow the pattern I observed in the existing code.